core/chain/security/blockaid/service.ts

```
@@ -0,0 +1,64 @@
+import { attempt } from '@lib/utils/attempt'
+
+import { queryBlockaid } from './blockaid'
+import { buildBlockaidScanPayload } from './buildScanPayload'
+import { BlockaidSupportedChains } from './buildScanPayload'
+import { Endpoints } from './constants'
+import { BlockaidErrorTypes, BlockaidResultTypes } from './constants'
+import { BlockaidError, BlockaidScanResult } from './types'
+
+export async function scanTransaction({
+  chain,
+  accountAddress,
+  rawTx,
+  metadata = { domain: 'https://vultisig.com' },
+}: {
+  chain: BlockaidSupportedChains
+  accountAddress?: string
+  rawTx?: string
+  metadata?: Record<string, unknown>
+}): Promise<BlockaidScanResult> {
+  const result = await attempt(async () => {
+    const payload = buildBlockaidScanPayload({
+      chain,
+      accountAddress,
+      rawTx,
+      metadata,
+    })
+    const endpoint = Endpoints.txScan(chain)
+    return await queryBlockaid(endpoint, payload)
+  })
+
+  if (!result.data) {
+    throw new Error('Blockaid scan returned no data')
+  }
+
+  return result.data
+}
+
+export function validateScanResult(scanResult: BlockaidScanResult): {
+  isValid: boolean
+  error?: BlockaidError
+} {
+  const validation = scanResult.validation
```

It shouldn't be exported and should be part of the `scanTransaction` function. We only care about the result: if it fails, we should throw an error; if it validates but is a bad transaction, we should return a validation result; and if everything is fine, we should return `undefined`. This way, we can have a single function, `validateTransaction`, that handles everything and returns what we truly care about: the result or `undefined` if everything is good.